crates/flipperzero/src/subghz/device.rs

```diff
@@ -0,0 +1,193 @@
+//! Safe wrapper utilizing the API provided in `lib/subghz/devices/devices.h`
+
+use super::{error::SubGhzError, preset::SubGhzPreset};
+use core::{ffi::CStr, ptr::null_mut};
+use flipperzero_sys::{
+    subghz_devices_begin, subghz_devices_deinit, subghz_devices_end, subghz_devices_flush_rx,
+    subghz_devices_flush_tx, subghz_devices_get_by_name, subghz_devices_idle, subghz_devices_init,
+    subghz_devices_is_connect, subghz_devices_is_frequency_valid,
+    subghz_devices_is_rx_data_crc_valid, subghz_devices_load_preset, subghz_devices_read_packet,
+    subghz_devices_reset, subghz_devices_rx_pipe_not_empty, subghz_devices_set_frequency,
+    subghz_devices_set_rx, subghz_devices_set_tx, subghz_devices_sleep,
+    subghz_devices_write_packet, SubGhzDevice,
+};
+use uom::si::{frequency::hertz, u32::Frequency};
+
+pub struct SubGhz {
+    device: &'static SubGhzDevice,
+}
+
+impl SubGhz {
+    /// For the internal device use "cc1101_int" (From `lib/subghz/devices/cc1101_int/cc1101_int_interconnect.c`)
+    pub fn subghz_devices_get_by_name(name: &CStr) -> Option<SubGhz> {
+        unsafe { subghz_devices_init() }
+
+        // Safety: input type enforeced by CStr type.
+        let dev = unsafe { subghz_devices_get_by_name(name.as_ptr()) };
+        // Safety: This pointer should either be null or return a reference, I don't see how it could not.
+        // Further, the reference should be static as it relates to the variable:
+        // `static SubGhzDeviceRegistry* subghz_device_registry = NULL;` in `lib/subghz/devices/registry.c``
+        let dev = unsafe { dev.as_ref()? };
+        Some(Self { device: dev })
+    }
+
+    // No clue what sort of error may occur, if using the internal device, it will always return Ok()
+    pub fn begin(&mut self) -> Result<(), SubGhzError> {
+        // Safety: self.device is not Null so this will not crash when furi_check is invoked.
```

Comment:
Important: `furi_check` causing a crash is *not* unsafe thus it should not be mentioned in safety comments (in the same matter as `panic!` is safe to call), although mentioning non-nullity is useful.

What should be mentioned here is if there are any `*const SubGhzDevice` invariants which should be upheld.

Same goes for the following `unsafe` calls in the file.

Reply:
I am aware that a crash is considered safe; as you mentioned I wanted to put it there to mention the non-nullity.

This is something I want to look into more, so thanks for your feedback on this. I will mark this comment as resolved as you have more details in your other comment.

Reply:
As a sidenote, I think I lost a bunch of replys to some of your comments since I clicked "Resolve Conversation" without adding the comment :/ I will go back and try add those comments, but most of them were just links to the commits in which I fixed them.